keras/layers/crf.py

```diff
@@ -0,0 +1,220 @@
+# -*- coding: utf-8 -*-
+from __future__ import absolute_import
+
+from .. import backend as K
+from .. import initializations, regularizers, constraints
+from ..engine import Layer, InputSpec
+
+
+def sparse_chain_crf_loss(y, x, U, b):
+    '''Given the true sparsely encoded tag sequences y, observations x,
+    transition params U and label bias b, it computes the loss function
+    of a Linear Chain Conditional Random Field.
+    '''
+    energy = path_energy(y, x, U, b)
+    energy -= free_energy(x, U, b)
+    return -energy
+
+
+def chain_crf_loss(y, x, U, b):
+    '''Given the true tag sequences y as one-hot encoded vectors,
+    observations x, transition params U and label bias b, it computes the loss
+    function of a Linear Chain Conditional Random Field.
+    '''
+    y_sparse = K.argmax(y, -1)
+    y_sparse = K.cast(y_sparse, 'int32')
+    return sparse_chain_crf_loss(y_sparse, x, U, b)
+
+
+def viterbi_decode(x, U, b):
+
+    def _forward_step(x_t, states):
+        gamma_tm1, alpha_tm1, U_shared = states
+        B = K.expand_dims(alpha_tm1, 2) + K.expand_dims(x_t, 1) + K.expand_dims(U_shared, 0)
+        alpha_t = K.max(B, axis=1)
+        gamma_t = K.cast(K.argmax(B, axis=1), K.floatx())
+        return gamma_t, [gamma_t, alpha_t]
+
+    inputs = x[:, 1:, :]
+    constants = [U]
+
+    alpha_0 = x[:, 0, :] + K.expand_dims(b, 0)
+    gamma_0 = K.zeros_like(alpha_0)
+    initial_states = [gamma_0, alpha_0]
+    _, gamma, states = K.rnn(_forward_step, inputs, initial_states, constants=constants)
+    gamma = K.cast(gamma, 'int32')
+    last_alpha = states[1]
+
+    def _backward_step(gamma_t, states):
+        beta_tm1 = states[0]
+        beta_t = K.batch_gather(gamma_t, beta_tm1)
+        return beta_t, [beta_t]
+
+    beta_m1 = K.cast(K.argmax(last_alpha, axis=1), 'int32')
+    initial_states = [beta_m1]
+    _, beta, _ = K.rnn(_backward_step, gamma, initial_states, go_backwards=True)
+    beta = K.reverse(beta, 1)
+    beta = K.permute_dimensions(beta, [1, 0])
+    best_sequence = K.concatenate([beta, K.expand_dims(beta_m1, 0)], axis=0)
+    best_sequence = K.permute_dimensions(best_sequence, [1, 0])
+    return best_sequence
+
+
+def free_energy(x, U, b):
+
+    def _forward_step(x_t, states):
+        alpha_tm1, U_shared = states
+        B = K.expand_dims(alpha_tm1, 2) + K.expand_dims(x_t, 1) + K.expand_dims(U_shared, 0)
+        alpha_t = K.logsumexp(B, axis=1)
```

Comment:
Thanks a lot for these remarks. You are completely right, that the models are not the same.

I don't handle batches of different length explicitely. They are embedded in sequences of fixed length by design, so the user has to do the conversion himself.